src/go/rpk/pkg/cli/topic/analyze.go

```diff
@@ -0,0 +1,743 @@
+		}
+
+		if len(batchSizes) != 0 {
+			pBatchSize, err := utils.Percentile(batchSizes, p)
+			if err != nil {
+				return summary, err
+			}
+			summary.BatchSizes = append(summary.BatchSizes, percentileValue[int]{p, pBatchSize})
+		} else {
+			summary.BatchSizes = append(summary.BatchSizes, percentileValue[int]{p, 0})
+		}
+	}
+
+	return summary, nil
+}
+
+type printConfig struct {
+	Percentiles        []float64
+	PrintGlobalSummary bool
+	PrintTopicSummary  bool
+}
+
+func (cr *collectResult) printSummaries(cfg printConfig) error {
+	percentiles := cfg.Percentiles
+	percentileNames := []string{}
+	for _, p := range percentiles {
+		percentileNames = append(percentileNames, fmt.Sprintf("P%v", p))
+	}
+
+	const (
+		secSummary   = "summary"
+		secBatchRate = "partition batch rate (1/s)"
+		secBatchSize = "partition batch size (bytes)"
+	)
+
+	ts, err := cr.summarizeTopics(percentiles)
+	if err != nil {
+		return err
+	}
+	// Print the topic summaries in asc order by topic name
+	slices.SortFunc(ts, func(a topicSummary, b topicSummary) int {
+		return strings.Compare(a.Topic, b.Topic)
+	})
+
+	gs, err := cr.summarizeGlobal([]float64{50})
+	if err != nil {
+		return err
+	}
+
+	sections := out.NewMaybeHeaderSections(
+		out.ConditionalSectionHeaders(map[string]bool{
+			secSummary:   cfg.PrintGlobalSummary,
+			secBatchRate: cfg.PrintTopicSummary,
+			secBatchSize: cfg.PrintTopicSummary,
+		})...,
+	)
+
+	sections.Add(secSummary, func() {
+		tw := out.NewTabWriter()
+		tw.PrintColumn("topics", gs.Topics)
+		tw.PrintColumn("partitions", gs.Partitions)
+		tw.PrintColumn("total batch rate (1/s)", gs.TotalBatchesPerSecond)
+		tw.PrintColumn("p50 batch size (bytes)", gs.BatchSizes[0].value)
+		tw.Flush()
+	})
+
+	sections.Add(secBatchRate, func() {
+		tw := out.NewTable(append([]string{"TOPIC"}, percentileNames...)...)
+		defer tw.Flush()
+		for _, ts := range ts {
+			vals := []interface{}{}
+			for _, pv := range ts.BatchesPerSecond {
+				vals = append(vals, pv.value)
+			}
+
+			tw.Print(append([]interface{}{ts.Topic}, vals...)...)
+		}
+	})
+
+	sections.Add(secBatchSize, func() {
+		tw := out.NewTable(append([]string{"TOPIC"}, percentileNames...)...)
+		defer tw.Flush()
+		for _, ts := range ts {
+			vals := []interface{}{}
+			for _, pv := range ts.BatchSizes {
+				vals = append(vals, pv.value)
+			}
+
+			tw.Print(append([]interface{}{ts.Topic}, vals...)...)
+		}
+	})
+	return nil
+}
+
+type kvTPInfo struct {
+	TopicPartition topicPartition      `json:"topicPartition" yaml:"topic_partition"`
+	Info           *topicPartitionInfo `json:"info" yaml:"info"`
+}
+
+type rawResults struct {
+	TopicPartitionInfo []kvTPInfo `json:"topicPartitionInfo" yaml:"topic_partition_info"`
+	TimeRange          timeRange  `json:"timeRange" yaml:"time_range"`
+}
+
+func (cr collectResult) printRawResults(f config.OutFormatter, w io.Writer) (bool, error) {
+	// Json doesn't seem to like serializing the map types. So convert it to a slice here.
```

Comment:
What was the error? Maps should be supported by JSON without any issues.

Perhaps there are missing struct tags in the map type definition? :thinking:

Reply:
With the current;
```go
type tpInfoMap map[topicPartition]*topicPartitionInfo
```
I get;

```
failed to print results: json: unsupported type: topic.tpInfoMap
```
And with;
```go
type tpInfoMap map[topicPartition]*topicPartitionInfo
```
I get;
```
failed to print results: json: unsupported type: map[topic.topicPartition]*topic.topicPartitionInfo
```
So the json serializer doesn't seem to like map types. Both the type definition and alias serialize with yaml just fine.

Reply:
Oh my bad, I got confused thinking it was a map of [string]any, but I'm seeing that you'll like to use tpInfoMap which is a `map[topicPartition]*topicPartitionInfo`, and `topicPartition` is a struct.

> JSON objects only support strings as keys; to encode a Go map type it must be of the form map[string]T (where T is any Go type supported by the json package). 

So, what you have is fine 😅. If you want to keep using the map type an alternative would be implementing the marshaller interface and define there how would you want to marshal the type (similar to what we did in https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/config/weak.go) but probably not worth it for this one.